examples/usecases/RAG_based_LLM_serving/README.md

```diff
@@ -0,0 +1,408 @@
+        self.vectorstore = FAISS.from_documents(
+            all_splits, CustomEmbedding(model_path=model_path)
+        )
+
+    def preprocess(self, requests):
+        inputs = []
+        for request in requests:
+            input_text = request.get("data") or request.get("body")
+            if isinstance(input_text, (bytes, bytearray)):
+                input_text = input_text.decode("utf-8")
+            inputs.append(input_text)
+        return inputs[0]
+
+    @torch.inference_mode
+    def inference(self, data, *args, **kwargs):
+        searchDocs = self.vectorstore.similarity_search(data, k=self.N)
+        return (searchDocs, data)
+
+    def postprocess(self, data):
+        docs, question = data[0], data[1]
+        doc_prompt = PromptTemplate.from_template("{page_content}")
+        context = ""
+        for _, doc in enumerate(docs):
+            context += f"\n{format_document(doc, doc_prompt)}\n"
+
+        prompt = (
+            f"Use the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer."
+            f"\n\n{context}\n\nQuestion: {question}"
+            f"\nHelpful Answer:"
+        )
+        return [prompt]
+```
+
+### Benchmarking Performance
+
+We use ab tool to measure the performance of the RAG endpoint
+
+```bash
+python benchmarks/auto_benchmark.py --input /home/ubuntu/serve/examples/usecases/RAG_based_LLM_serving
+benchmark_profile.yaml --skip true
+```
+
+
+We repeat the runs with combinations of OMP_NUM_THREADS and PyTorch Eager/ torch.compile
+
+
+#### Results
+
+We observe the following throughput on the AWS EC2  `m7g.4xlarge` instance
+
+![RAG Throughput](https://raw.githubusercontent.com/pytorch/serve/master/examples/usecases/RAG_based_LLM_serving/assets/rag_perf.png "RAG Throughput")
+
+
+We observe that using `torch.compile` improves the RAG endpoint throughput by 35%. The scale of the throughput (Eager or Compile) shows that deploying RAG on a CPU device is practical for use with a LLM deployed on a GPU instance. The RAG endpoint is not going to be a bottleneck in an LLM deployment,
+
+
+## RAG + LLM Deployment
+
+The system architecture for the end-to-end solution using RAG based LLM serving is shown in Figure 2.
+
+
+
+![RAG + LLM Deployment](https://raw.githubusercontent.com/pytorch/serve/master/examples/usecases/RAG_based_LLM_serving/assets/rag_llm.png "RAG + LLM Deployment")
+
+
+The steps for full deployment are mentioned in [Deploy.md](https://github.com/pytorch/serve/blob/master/examples/usecases/RAG_based_LLM_serving/Deploy.md)
+
+The code snippet which can chain the RAG endpoint with Llama endpoint is shown below
+
+
+```python
+import requests
+
+prompt="What's new with Llama 3.1?"
+
+RAG_EP = "http://<RAG Endpoint IP Address>:8080/predictions/rag"
+LLAMA_EP = "http://<RAG Endpoint IP Address>:8080/predictions/llama3"
```

Comment:
model name will be **llama3-8b-instruct**

Reply:
done